### PR TITLE
refactor(tiny_fft): refactor tiny_ttf cache

### PR DESCRIPTION
**Pull request by W-Mai**
### Description of the feature or fix

Tiny_ttf has been reconstructed:
1. cache of uincode(letter) -> glyph_dsc  @codewitch-honey-crisis @MacDragon
2. lv_tiny_ttf_create_xxxx_ex really affect the cache size @cybekRT
3. all tiny_ttf use a same cache -> each font use separate caches  

Theoretically, it accelerates the creation speed of tiny_ttf text.

- https://github.com/lvgl/lvgl/issues/4835#issuecomment-2134459861
- #6305 
- #6271 

cc @kisvegabor 

Could you @MacDragon help test whether there is a performance improvement?

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.12](https://github.com/szepeviktor/astyle/releases/tag/v3.4.12) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by MacDragon** [comment]
I'll check it when current time critical stuff is done, hopefully friday. I'll see if I can import latest master tree into a working branch to run comparisons on, as @codewitch-honey-crisis mentioned that there was already noticeable rendering speed improvements without since 9.1 release due to other changes it's probably best to try and run the different options against newer tree.

**Comment by W-Mai** [comment]
It's a strange bug. Why do the coordinates move up a pixel? 🤔

**Comment by W-Mai** [comment]
fixed, it's stupid mistake 😄💦

image

**Comment by C47D** [comment]
> fixed, it's a stupid mistake 😄💦
> 
> image
> 
> 

Nice font, which is it?

**Comment by W-Mai** [comment]
> > fixed, it's a stupid mistake 😄💦
> > image
> 
> Nice catch and also nice font, which is it?

Pixeloid Mono 😁
I'm glad you also like this font.

- https://www.dafont.com/pixeloid-mono.font

File: src/libs/tiny_ttf/lv_tiny_ttf.h [hunk 1]
```diff
@@ -37,10 +37,6 @@ lv_font_t * lv_tiny_ttf_create_file(const char * path, int32_t font_size);
 lv_font_t * lv_tiny_ttf_create_file_ex(const char * path, int32_t font_size, size_t cache_size);
```
Comment (kisvegabor) [inline src/libs/tiny_ttf/lv_tiny_ttf.h:37]: Please fix the API comments of these functions.

**Review by kisvegabor — COMMENTED** [review]
Thanks @W-Mai! It looks good in general. I cannot test the performance difference  now, but if someone verifies it, we are good to go. 

Just one comment:

**Comment by kisvegabor** [comment]
Could you also take care of this?
https://github.com/lvgl/lvgl/issues/6308

**Comment by kisvegabor** [comment]
And a good catch here: #6340

**Comment by MacDragon** [comment]
Today ended up a fair bit busier than I had anticipated, but monday I should  be clear to do some comparison trialing.

**Comment by MacDragon** [comment]
Need to do a little more testing, but currently I do not appear to see any difference between master branch and applying the cache update atop:

Master branch:

 24.06.10 12:02.42      4.033 flush done  5:38 -> 315:190 81us 47583
 24.06.10 12:02.42      4.034 render done 284942us
redraw:
 24.06.10 12:05.28    170.187 flush done  5:38 -> 315:190 93us 47583
 24.06.10 12:05.28    170.187 render done 263750us
 
 
Cache branch:
 24.06.10 12:13.26      3.276 flush done  5:38 -> 315:190 82us 47583
 24.06.10 12:13.26      3.276 render done 286044us
 redraw:
 24.06.10 12:13.52     29.059 flush done  5:38 -> 315:190 93us 47583
 24.06.10 12:13.52     29.059 render done 264225us

Need to verify it's actually active and then do another comparison tomorrow morning.

**Comment by MacDragon** [comment]
Ok, I'm further investigating this today, I think I possibly have some improvements - I'm first fixing up my hacky implementation to be more correct, and then seeing if I can apply some fixes here and see if they help.

**Comment by kisvegabor** [comment]
Great, thank you!

**Comment by MacDragon** [comment]
still investigating, it looks like the really expensive part of the call for me is actually the kerning lookup, and my original implementation had completely ignored this as unecessary in my use case - as I was using a monospaced font, so there should be no kerning.

It might be that need an option to disable kerning on font creation, when know don't need it - kerning lookups can probably also be cached in some form too.

**Comment by codewitch-honey-crisis** [comment]
I'd just like to add here, if it's helpful, that anything Tiny TTF has to do a table lookup for is going to be necessarily somewhat slow, and slower the larger the table. The reason being is that it has to do a lookup over the TTF/OTF file as it exists on disk/flash. It puts memory at a premium and so it does not create indexed lookup tables for this stuff although it will use a binary search if the TTF/OTF file keeps a particular table in sorted order. Might be helpful to bear in mind, as it affects kerning lookup, glyph lookup, etc. This is actually I think the best case scenario, because we're effectively punting faster lookups to the cache which is better managed than if TinyTTF was creating those tables itself. Also @MacDragon thanks for your help on this. I simply don't have time presently to make improvements to the code, so I appreciate all your hard work.

**Comment by codewitch-honey-crisis** [comment]
After thinking about your comment some more @MacDragon it occurs to me that you can't actually technically disable the kerning for a font from a TTF/OTF even if monospaced. The issue is that each glyph has an "advance" property that tells the font how many pixels to advance along the X axis. That's a per glyph lookup even for monospaced fonts. At best, you could cache that for monospaced fonts, but frankly, I think it would be easier on the user and more flexible to integrate the advance factor into the cache.

**Comment by MacDragon** [comment]
Yes, that's effectively what I'm looking to do, cache it and use default spacing for char if no next char or kerning flagged off.

**Comment by MacDragon** [comment]
Ok, in my current case rendering has now improved to:

24.06.12 12:01.31     10.702 flush done  5:38 -> 315:190 142us 47583
 24.06.12 12:01.31     10.702 render done 10806us

Though this is a pretty idealistic state with I think effectively no kerning really happening - need to find some test case that actually triggers real kerning need. But yes, with a few tweaks this is already getting much better now for my present use case at least.

**Comment by W-Mai** [comment]
> Ok, in my current case rendering has now improved to:
> 
> 24.06.12 12:01.31 10.702 flush done 5:38 -> 315:190 142us 47583 24.06.12 12:01.31 10.702 render done 10806us
> 
> Though this is a pretty idealistic state with I think effectively no kerning really happening - need to find some test case that actually triggers real kerning need. But yes, with a few tweaks this is already getting much better now for my present use case at least.

Although kerning cannot be completely banned, can you also test the code of the cache branch to remove the calculation of kerning, will it reduce the time consumption?

**Comment by MacDragon** [comment]
Yes, there is definitely gains to be had by a few less floating point calculations.

**Comment by MacDragon** [comment]
With kerning on, but no actual kerning required to be.
 24.06.12 13:03.12      2.664 flush done  5:38 -> 315:190 130us 47583
 24.06.12 13:03.12      2.664 render done 16588us
after cached and redrawn:
 24.06.12 13:03.23     13.459 flush done  5:38 -> 315:190 131us 47583
 24.06.12 13:03.23     13.459 render done 9366us

In current test I seem to be seeing effectively no difference between kerning on and off - though I don't think the fon't I'm testing with ever triggers real kerning so I think it's a bad test case.. I'll try to find some case tomorrow morning that does, but optimising by adding no kerning calculation to the cache creation seems to have cut another 1ms off my case after cacheing done, though this seems to be about the same whether kerning enabled or not. I'll continue tomorrow morning for a bit and try and find a different test case.

**Comment by MacDragon** [comment]
Also of note, I am testing on an MPU with hardware floating point, the gains may well be more notable on software only. Once I've got something that seems reasonable to test further what's best way to submit?

**Comment by MacDragon** [comment]
Currently working on this again, I don't seem to be reliably getting unicode_letter_next for kerning checks in lv_label?

**Comment by MacDragon** [comment]
Ok, now I think the kerning enable is working properly:
no kern first draw:

 24.06.13 08:12.28      6.428 flush done  5:38 -> 315:190 131us 47583
 24.06.13 08:12.28      6.428 render done 19351us

kern first draw: 
24.06.13 08:11.28      5.533 flush done  5:38 -> 315:190 94us 47583
 24.06.13 08:11.28      5.533 render done 238175us

Kern pairing data should still probably be cached in some way also, but even so the kerned version is still running in 2/3 the time it used to be.

**Comment by MacDragon** [comment]
Yes, that helped quite a bit more:

 24.06.13 08:44.30      5.356 flush done  5:38 -> 315:190 87us 47583
 24.06.13 08:44.30      5.356 render done 143749us

I think this is pro

**Comment by MacDragon** [comment]
[https://github.com/MacDragon/lvgl/tree/%F0%9F%90%BB](url) Has what I've currentlly been trialling to look at.

It should disable kerning automatically if no tables are found, _ex has option to disable on request - which speeds it up a lot if don't need it. Otherwise I'm sure a better way of cacheing the kern pairs could still be found, it didn't seem to speed up a lot when I changed the cache size so I'm not sure if it's working right or not for sure.

**Comment by kisvegabor** [comment]
I've tested your changes on PC and it `kern caching` really makes a difference! I can test it on embedded hardware on Monday. Until that @W-Mai could you review the changes of @MacDragon and if they look good integrate into this PR?

**Comment by MacDragon** [comment]
Taken a little more look, and I'm not sure if the actual kerning lookups are working properly.. so I definitely wouldn't pull quite yet, I'm investigating a bit further. - I only seem to be getting 0 returns on copy of helvetica I was just now testing with.

**Comment by MacDragon** [comment]
Yes, it looks like the fonts I've tried support a newer kerning table type than tiny ttf currently supports, hence the lookup time was constant.. I'm adding a check for invalid table type to save pointless lookup, and trying to find an actually supported font to verify and test.

**Comment by MacDragon** [comment]
Ok, ixed it, so the current version actually works - and now disables kerning if incompatible table format.

Will do performance testing a bit later.

**Comment by MacDragon** [comment]
At first glance, performance appears to be excellent now - with actual working kerning cache.

**Comment by MacDragon** [comment]
Another thought occurs, technically the kerning cache could be shared between different instances of same font to save memory, not sure how tricky that would be to implement.

**Comment by kisvegabor** [comment]
I've just tested @MacDragon's version.

**Ubuntu font** (Kerning is not supported):
- Kerning: None
  - Cache size: 1
     - Render time: 195ms
   - Cache size: 256
      - Render time: 39ms
- Kerning: Normal
  - No change


**Arial font** (Kerning is supported)
  - No kerning: 40ms
  - Kern cache 0:   50ms
  - Kern cache 1:   87ms
  - Kern cache 256: 52ms
  - Kern cache 4096: 51ms

Kern cache 0 (no cache) is very strange. Why is it faster than cache 1?

**Comment by MacDragon** [comment]
Hmm, that does indeed sound odd - is it actually skipping the kerning lookups on 0 and just always assuming 0?

**Comment by kisvegabor** [comment]
With 0 kern cache size it used `stbtt_GetGlyphKernAdvance` which returned `-186`. 
I've used this font: [Arialn.ttf.zip](https://github.com/user-attachments/files/15904368/Arialn.ttf.zip)

**Comment by MacDragon** [comment]
True, it should be running the command directly if cache lookup fails which with 0 cache it should I suppose.. I'll see if I can run more few more tests of own for comparison, it certainly sped up my original test rendering when I added the kern cache vs only glyph desc lookups.

**Comment by MacDragon** [comment]
Finally got some time to look at this again, and I can confirm your odd looking results appear to replicate here, will investigate further as something obviously not going as expected.

    //testfont = lv_tiny_ttf_create_data_ex2(Arialn_ttf, sizeof Arialn_ttf, 16, LV_FONT_KERNING_NORMAL, 1024, 0);
    //                    1.368 render done 14223us
    //                    2.369 render done 7238us

    //testfont = lv_tiny_ttf_create_data_ex2(Arialn_ttf, sizeof Arialn_ttf, 16, LV_FONT_KERNING_NORMAL, 1024, 1);
    //                    1.375 render done 26620us
    //                    2.362 render done 11299us

    //testfont = lv_tiny_ttf_create_data_ex2(Arialn_ttf, sizeof Arialn_ttf, 16, LV_FONT_KERNING_NORMAL, 1024, 256);
    //                    1.360 render done 15643us
    //                    2.360 render done 7695us

first number is first render, second number is second cached render. They appear to bear out your observations that 0 is somehow faster.

**Comment by MacDragon** [comment]
I think it may even be that the actual kerning lookup is slower cached, and most of the speedup was the second glyph cache lookup. Which would actually be good as that would save extra cache - I implemented both parts at same time with assumption it would be faster.

**Comment by MacDragon** [comment]
Yes, it looks like the cache overhead is bigger than the direct lookup speed in normal cases.

I suspect it would be more beneficial in the newer more complex kerning lookups - but it looks like it is indeed best to at least default kern cacheing itself to zero for now, and only use the glyph lookups.

**Comment by MacDragon** [comment]
Tidied away the kern cache for now and cleaned up formatting so it passes verifier:

https://github.com/MacDragon/lvgl/tree/tiny_ttf_cache

**Comment by kisvegabor** [comment]
Awesome! @MacDragon as you version is a superset of this one, could you open a new PR?

**Comment by kisvegabor** [comment]
Let's continue in #6441

**Pull request closed without merging**